Placeholder for JobSuccessPolicy feature blog (#49998)
kubernetes/website: content/en/blog/_posts/2025-04-23-jobs-successpolicy-goes-ga.md

---
layout: blog
title: "Kubernetes 1.33: Job's SuccessPolicy Goes GA"
date: 2025-04-23
slug: kubernetes-1-33-jobs-success-policy-goes-ga
authors: >
  [Yuki Iwai](https://github.com/tenzen-y) (CyberAgent, Inc)
---

> Announcing the graduation to General Availability (GA) of the Job's __Success Policy__, in Kubernetes v1.33.

Reviewer: ```suggestion
We're pleased to announce that Job's __Success Policy__ will  graduate to General Availability (GA) as part of Kubernetes v1.33 release.
```

Author: That sounds more clarifying; thank you!

## About Job's Success Policy

> When you run batch workloads with leader-worker patterns, you might want to mark it as succeeded

Reviewer: If you can point to some examples, wiki page it'd be nice. Also the majority of today's literature is using leader-follower name, so I'd suggest using that here as well.

Author: Thank you. I will replace `leader-worker` with `leader-follower` in all places.
Additionally, I am trying to explore the wiki for leader-follower pattern.

> even if some index failed. On the other hand, the Kubernetes Job without __Success Policy__ requires all Pods succeeded for Job succeeded.

Author: Thank you for sharing the context. That makes sense. Let me apply your recommendations to this PR.

> even if some the indexes failed failed. Unfortunately, the Kubernetes Job without __Success Policy__ requires in most cases all Pods succeeded for Job succeeded.

Author: Thanks

> This Job's __Success Policy__ allows you to specify what is criteria for Job succeeded under the `.spec.successPolicy` 
> opposed to the default success criteria which is all indexes succeeded.

Reviewer: ```suggestion
Job's __Success Policy__ allows you to specify the early exit criteria for a Job using  the `.spec.successPolicy` 
field. Which describes a set of rules either using a list of succeeded indexes for a job, or defining a minimal required size of succeeded indexes. 
```

Author: That makes sense since your recommendation describes actual new fields, thanks

Author: > It might be good to preface this with a statement that this feature is only available for indexed jobs.

That makes sense.

> We assume this feature is useful for the scientific simulation workloads and AI/ML or High-Performance computing batch workloads.

Reviewer: ```suggestion
This feature is especially useful for the scientific simulation workloads and AI/ML or High-Performance computing batch workloads.
```

> The scientific workloads sometimes perform simulations with different parameters and do not want to care about the number of successful experiments.

Reviewer 2: +1, I think what is meant here is that the logic determining if the Job is successful or not is in leader, and the Job controller does not need to inspect or count failures of workers directly. I would suggest to consider adding an example for this use-case, or probably just editing the second example from rather abstract `succeededIndexes: 0,2-4`, to `succeededIndexes: 0`. Indicating, that the fate of the Job is determined just based on leader index.

> The scientific workloads sometimes perform simulations with different parameters and do not want to care about the number of successful experiments.

Author: I thought `0,2-4` indicates a HighAvailability leader. However, in actual use cases like MPI, they support only a single leader. So, I replaced `0,2-4` with `0`. Thank you.

> The scientific workloads sometimes perform simulations with different parameters and do not want to care about the number of successful experiments.

Reviewer 2: thanks

> The AI/ML or High-Performance computing workloads sometimes use leader-worker patterns like MPI, in which the leader controls everything, including the workers' lifecycles.

Reviewer: ```suggestion
The AI/ML or High-Performance computing workloads sometimes use leader-follower patterns like MPI, in which the leader controls everything, including the followers' lifecycle.
```

Author: SGTM

In this case, the Job failure criteria are leader index failure, and workers failure do not indicate Job failure.

> Additionally, workers do not know when they can terminate themselves.

Author: SGTM, thank you

> In this case, the Job failure criteria are leader index failure, and workers failure do not indicate Job failure.

Reviewer: ```suggestion
In this case, the leader index failure is the only the Job's failure criteria, and workers results is not relevant.
```

Author: Thanks

> After Job meets any __Success Policy__, the Job goes to succeeded state and all Pods are terminated including running Pods.

Reviewer: ```suggestion
After Job meets any __Success Policy__, the Job is marked as succeeded, and all Pods are terminated including running ones.
```

> Additionally, workers do not know when they can terminate themselves.

Reviewer: ```suggestion
Moreover, workers do not know when they may terminate themselves.
```

Author: Thanks

> After Job meets any __Success Policy__, the Job is marked as succeeded, and all Pods are terminated including running ones.

Reviewer: ```suggestion
After Job meets any __Success Policy__, the Job is marked as succeeded, and all Pods are terminated including the running ones.
```

## How it works

> The below Job spec with `.successPolicy.rules[0].succeededCount` shows the example of __Success Policy__ feature:

Reviewer: ```suggestion
The below Job spec with `.successPolicy.rules[0].succeededCount` shows an example of __Success Policy__ feature:
```

```yaml
parallelism: 10
completions: 10
completionMode: Indexed
successPolicy:
  rules:
  - succeededCount: 1
```

> In this example, the Job is marked as succeeded when one index succeeded regardless of index number.

Reviewer: ```suggestion
Here, the Job is marked as succeeded when one index succeeded regardless of its number.
```

Additionally, we can add constraints for index numbers against `succeededCount` to `.successPolicy.rules[0].succeededCount`
as shown below:

```yaml
parallelism: 10
completions: 10
completionMode: Indexed
successPolicy:
  rules:

>   - succeededIndexes: 0,2-4

Reviewer 2: This example looks rather abstract, maybe we can consider updating it to just say `succeededIndexes: 0` which would better reflect the use case of determining the fate of the Job just based on the status of the leader index. Also proposed in https://github.com/kubernetes/website/pull/49998/files#r2041573529

Author: SGTM, I applyied the recommendation to this example, thanks.

Reviewer 2: thanks

    succeededCount: 1
```

> This example shows that the Job is marked as succeeded when one of the indexes 0, 2, 3, or 4 succeeded.

Reviewer 3: It might be also a good chance to explain the new `SuccessCriteriaMet` condition.

Reviewer 3: Btw, I have just spot a mistake by an accident. https://github.com/kubernetes/kubernetes/blob/b15dfce6cbd0d5bbbcd6172cf7e2082f4d31055e/pkg/apis/batch/types.go#L263 references `SucceededCriteriaMet`, but it should say `SuccessCriteriaMet` instead.

Reviewer: > It might be also a good chance to explain the new `SuccessCriteriaMet` condition.

While re-reading I was thinking about the exactly the same thing. It's definitely a crucial concept to fully understand how a job reaches the early success criteria, even though it's not fully finished.

Author: > Btw, I have just spot a mistake by an accident. https://github.com/kubernetes/kubernetes/blob/b15dfce6cbd0d5bbbcd6172cf7e2082f4d31055e/pkg/apis/batch/types.go#L263 references SucceededCriteriaMet, but it should say SuccessCriteriaMet instead.

@atiratree Oh, you're good eye. Thank you for mentioning the typo. I will open a fixing PR.

Author: > > It might be also a good chance to explain the new SuccessCriteriaMet condition.
> 
> While re-reading I was thinking about the exactly the same thing. It's definitely a crucial concept to fully understand how a job reaches the early success criteria, even though it's not fully finished.

That sounds reasonable. Let me add description.

Reviewer 3: https://github.com/kubernetes/website/pull/49998#discussion_r2042578882

## About Job's Success Policy

In batch workloads, you might want to use leader-follower patterns like [MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface),
in which the leader controls the execution, including the followers' lifecycle.

In this case, you might want to mark it as succeeded
even if some the indexes failed. Unfortunately, the Kubernetes Job without __Success Policy__, in most cases,
requires all Pods to finish successfully for a Job to reach succeeded state.

Job's __Success Policy__ allows you to specify the early exit criteria using the `.spec.successPolicy` 
field only for the [Indexed Job](/docs/concept/workloads/controllers/job/#completion-mode). 
Which describes a set of rules either using a list of succeeded indexes for a job, or defining a minimal required size of succeeded indexes. 

This feature is especially useful for the scientific simulation workloads, AI/ML and High-Performance computing batch workloads. 
In those workloads users perform various simulations and might not necessarily care about all the experiments to complete successfully,
rather they require only a certain number of them to finish successfully. 

> In this case, the leader index failure is only the Job's failure criteria, and workers results is not relevant.

Reviewer 2: ```suggestion
In this case, the leader index failure is only the Job's failure criteria, and workers results are relevant
only indirectly via the status of the leader index.
```
I think saying "workers results is not relevant." may be a bit misleading. They are relevant, but indirectly. wdyt?

Author: That makes sense. Actually, I imagined the followers perform only sshd daemon. So, I said, `It is not relevant. '
However, the failure of sshd launching might indirectory relevant the JOB failure.

## About Job's Success Policy

In batch workloads, you might want to use leader-follower patterns like [MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface),
in which the leader controls the execution, including the followers' lifecycle.

In this case, you might want to mark it as succeeded
even if some the indexes failed. Unfortunately, the Kubernetes Job without __Success Policy__, in most cases,
requires all Pods to finish successfully for a Job to reach succeeded state.

Job's __Success Policy__ allows you to specify the early exit criteria using the `.spec.successPolicy` 
field only for the [Indexed Job](/docs/concept/workloads/controllers/job/#completion-mode). 
Which describes a set of rules either using a list of succeeded indexes for a job, or defining a minimal required size of succeeded indexes. 

This feature is especially useful for the scientific simulation workloads, AI/ML and High-Performance computing batch workloads. 
In those workloads users perform various simulations and might not necessarily care about all the experiments to complete successfully,

> rather they require only a certain number of them to finish successfully. 

Reviewer 3: ```suggestion
```

This is very similar to the first part of the sentence.

Author: Couldn't we consider this as rephrasing and clarification against the first part of the sentence?
Do you feel redundant?

Reviewer 3: https://github.com/kubernetes/website/pull/49998#discussion_r2042647408 solves this in a better way

In this case, the leader index failure is only the Job's failure criteria, and followers results are relevant
only indirectly via the status of the leader index.
Moreover, followers do not know when they can terminate themselves.

After Job meets any __Success Policy__, the Job is marked as succeeded, and all Pods are terminated including the running ones.

## How it works

The below Job spec with `.successPolicy.rules[0].succeededCount` shows an example of __Success Policy__ feature:

```yaml
parallelism: 10
completions: 10
completionMode: Indexed
successPolicy:
  rules:
  - succeededCount: 1
```

Here, the Job is marked as succeeded when one index succeeded regardless of its number.
Additionally, we can constrain index numbers against `succeededCount` in `.successPolicy.rules[0].succeededCount`
as shown below:

```yaml
parallelism: 10
completions: 10
completionMode: Indexed
successPolicy:
  rules:
  - succeededIndexes: 0
    succeededCount: 1
```

This example shows that the Job is marked as succeeded when one of the indexes 0 succeeded.

> Once the Job reaches the one of successPolicy rules or `Complete` criteria like `.spec.backoffLimit`,

Author: Ohh, you're right. Thank you for this great catch!

> Once the Job reaches the one of successPolicy rules or `Complete` criteria like `.spec.backoffLimit`,

Author: done.

the kube-controller-manager job-controller adds the `SuccessCriteriaMet` condition to the Job condition.

> After that, the job-controller executes terminating processes like terminating running Pods against Jobs with `SuccessCriteriaMet` condition. 

Author: That sounds reasonable, thanks.

> After that, the job-controller executes terminating processes like terminating running Pods against Jobs with `SuccessCriteriaMet` condition. 

Reviewer 3: Looks better, but maybe we can do something like https://github.com/kubernetes/website/pull/49998#discussion_r2042694651 and update the 2nd sentence as well?

> After that, the job-controller executes terminating processes like terminating running Pods against Jobs with `SuccessCriteriaMet` condition. 

Author: Thank you for pointing it out.
I will address it.

---
layout: blog
title: "Kubernetes 1.33: Job's SuccessPolicy Goes GA"

> date: 2025-04-23

Reviewer 4: While this blog is about the v1.33 release, we will limit to one blog published per day. This means it is likely that this blog will be planned on a later date, and the schedule is yet to be fully decided. (It is highly dependent on other reviews being complete in)

If you can simply add `draft: true` in the front matter, it would allow us to merge the blog without actually publishing them to the website. Once the blog content is finalised and merged, the Release Comms team can take over the rest of the publication setup.

```suggestion
date: 2025-04-23
draft: true
```

You can find more in the blog guideline [here](https://kubernetes.io/docs/contribute/blog/article-submission/#article-scheduling)! 👍

draft: true
slug: kubernetes-1-33-jobs-success-policy-goes-ga
authors: >
  [Yuki Iwai](https://github.com/tenzen-y) (CyberAgent, Inc)
---

> We're pleased to announce that Job's __Success Policy__ will  graduate to General Availability (GA) as part of Kubernetes v1.33 release.

Author: Thank you for your kind guidance. I will apply it everywhere in this PR.

> We're pleased to announce that Job's __Success Policy__ will  graduate to General Availability (GA) as part of Kubernetes v1.33 release.

Author: Done.

## About Job's Success Policy

In batch workloads, you might want to use leader-follower patterns like [MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface),
in which the leader controls the execution, including the followers' lifecycle.

In this case, you might want to mark it as succeeded

> even if some of the indexes failed. Unfortunately, the Kubernetes Job without __Success Policy__, in most cases,
> requires all Pods to finish successfully for a Job to reach succeeded state.

Reviewer 5: Maybe this:
```suggestion
even if some of the indexes failed. Unfortunately, a leader-follower Kubernetes Job that didn't use a success policy, in most cases, would have to require **all** Pods to finish successfully
for that Job to reach an overall succeeded state.
```
?

Job's __Success Policy__ allows you to specify the early exit criteria using the `.spec.successPolicy` 

> field (only for the [Indexed Job](/docs/concept/workloads/controllers/job/#completion-mode)). 

Reviewer 5: ```suggestion
field (you can only use the `.spec.successPolicy`  field for an [indexed Job](/docs/concept/workloads/controllers/job/#completion-mode)).
```

Which describes a set of rules either using a list of succeeded indexes for a job, or defining a minimal required size of succeeded indexes. 

> This feature is especially valuable for scientific simulation, AI/ML and High-Performance Computing (HPC) batch workloads. 

Reviewer 5: ```suggestion
This newly stable field is especially valuable for scientific simulation, AI/ML and High-Performance Computing (HPC) batch workloads.
```

Users in these areas often run numerous experiments and may only need a specific number to complete successfully, rather than requiring all of them to succeed. 

> In this case, the leader index failure is the only Job exit criteria, and followers results are relevant

Reviewer 5: ```suggestion
In this case, the leader index failure is the only relevant Job exit criteria, and the outcomes for individual follower Pods are handled
```

only indirectly via the status of the leader index.
Moreover, followers do not know when they can terminate themselves.

> After Job meets any __Success Policy__, the Job is marked as succeeded, and all Pods are terminated including the running ones.

Author: > its overall success policy

Actually, once one of success policies is satisfied, the Job is marked as succeeded.

> After Job meets any __Success Policy__, the Job is marked as succeeded, and all Pods are terminated including the running ones.

Author: I dropped `overall` like this:

> After a Job meets its success policy, that Job is marked as _succeeded_, and all Pods for that Job are terminated (even including the running ones).

## How it works

The below Job spec with `.successPolicy.rules[0].succeededCount` shows an example of __Success Policy__ feature:

```yaml
parallelism: 10
completions: 10
completionMode: Indexed
successPolicy:
  rules:
  - succeededCount: 1
```

Here, the Job is marked as succeeded when one index succeeded regardless of its number.

> Additionally, we can constrain index numbers against `succeededCount` in `.successPolicy.rules[0].succeededCount`

Author: Done.

as shown below:

```yaml
parallelism: 10
completions: 10
completionMode: Indexed
successPolicy:
  rules:

>   - succeededIndexes: 0

Reviewer 5: ```suggestion
  - succeededIndexes: 0 # index of the leader Pod
```

    succeededCount: 1
```

> This example shows that the Job is marked as succeeded when one of the indexes; index 0 succeeded.

Reviewer 5: ```suggestion
This example shows that the Job will be marked as succeeded once a Pod with a specific index (Pod index 0) has succeeded.
```

Once the Job reaches one of the successPolicy rules or `Complete` criteria based on `.spec.completions`,
the kube-controller-manager job-controller adds the `SuccessCriteriaMet` condition to the Job condition.
After that, the job-controller initiates cleanup and termination of Pods for Jobs with `SuccessCriteriaMet` condition.
Eventually, Jobs obtain `Complete` condition when the job-controller finished cleanup and termination.

## Learn more

- Read the documentation for
  [Success Policy](/docs/concepts/workloads/controllers/job/#success-policy)
- Read the KEP for the [Job success/completion policy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3998-job-success-completion-policy)

## Get involved

This work was sponsored by
[batch working group](https://github.com/kubernetes/community/tree/master/wg-batch)
in close collaboration with the
[SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps) community.

If you are interested in working on new features in the space we recommend
subscribing to our [Slack](https://kubernetes.slack.com/messages/wg-batch)

> channel and attending the regular community meetings.

Author: Done.

rather they require only a certain number of them to finish successfully. 
In this case, the leader index failure is the only Job failure criteria, and followers results are relevant
only indirectly via the status of the leader index.
Moreover, followers do not know when they can terminate themselves.

After Job meets any __Success Policy__, the Job is marked as succeeded, and all Pods are terminated including the running ones.

## How it works

The below Job spec with `.successPolicy.rules[0].succeededCount` shows an example of __Success Policy__ feature:

```yaml
parallelism: 10
completions: 10
completionMode: Indexed
successPolicy:
  rules:
  - succeededCount: 1
```

Here, the Job is marked as succeeded when one index succeeded regardless of its number.
Additionally, we can constrain index numbers against `succeededCount` in `.successPolicy.rules[0].succeededCount`
as shown below:

```yaml
parallelism: 10
completions: 10
completionMode: Indexed
successPolicy:
  rules:
  - succeededIndexes: 0
    succeededCount: 1
```

This example shows that the Job is marked as succeeded when one of the indexes; index 0 succeeded.

> Once the Job reaches the one of successPolicy rules or `Complete` criteria based on `.spec.completions`,

Reviewer 3: ```suggestion
Once the Job reaches one of the successPolicy rules or `Complete` criteria based on `.spec.completions`,
```

the kube-controller-manager job-controller adds the `SuccessCriteriaMet` condition to the Job condition.

> After that, the job-controller executes terminating processes like terminating running Pods against Jobs with `SuccessCriteriaMet` condition.

Reviewer 3: ```suggestion
After that, the job-controller initiates cleanup and termination of Pods for Jobs with `SuccessCriteriaMet` condition.
```
